### ISSUE TITLE
Petition to remove @apply support

### ISSUE THREAD

**Issue by ruphin** [Polymer/lit-html#370]
I noticed that support for CSS `@apply` had been added. I'm sure you are aware that `@apply` has been deprecated: https://www.chromestatus.com/feature/5753701012602880

If the purpose of this library is to be a low-level un-opinionated rendering engine, adding support for non-standard features is a terrible idea.

First, by adding explicit support for `@apply` you are signaling users that it is "okay" to use this feature. Since we already know that `@apply` _will_ no longer be used at some point, you are just causing migration problems for these users down the road. Not everyone is familiar with the platform status of certain features, and it is irresponsible to signal that features are good to use when they are in fact deprecated.

Furthermore, this feature does not even work in any browser without polyfills, and this will never change. Adding a feature that only ever works when polyfills are used is ridiculous. You're effectively adding weight to everyone's bundle for nothing, unless your intent is for the majority of users of this library to ship the `@apply` polyfill to all their users, which is also quite ridiculous.

I am very fond of lit-html and I think it has a great future as a rendering engine in many web based projects (we are already seeing this happen), but this comes with great responsibility. Please follow the standards.

**Comment by AndreasGalster** [comment]
If you remove lit-html basically, it is impossible to migrate to lit-html. We currently are blocked on using lit further due to shady CSS issues. Our app is 3.0 with some components migrated to lit but not all

**Comment by Lodin** [comment]
AFAIK, `lit-html` doesn't support `@apply` directly. It supports Shady CSS, which still supports `@apply`. So I suppose this issue has a wrong destination.

I also think that since `:: part`/`:: theme` standard is not fully established and still doesn't have proper polyfill, it's good to support `@apply` at least. Styling with only CSS Variables it's just a nightmare.

**Comment by ruphin** [comment]
How is it impossible? There are plenty of methods for shared CSS with lit-html that are standards compliant. It's not like it's impossible for web applications to work without `@apply`.

It's also basically impossible to migrate my React.js app to lit-html, so should it support JSX now?

If lit-html is supposed to be an independent project that serves as the render engine for _next generation_ components, that can be used by `lit-element` or any of the many tens of projects that are using it for this purpose, it should not carry any baggage from older projects.

I understand that it lives under the Polymer umbrella, and for that reason it's appealing to add features to ease migration from Polymer projects, but it can be so much more than that. I think it's a great foundation for any Web Component project, and could possibly even replace Virtual DOM in other frameworks.

If the intention of this project is purely to provide a stepping stone for legacy Polymer projects and disregard web standards, I feel like that eliminates it as a viable choice for other frameworks and projects to use, which is a real shame.

**Comment by ruphin** [comment]
Upon reviewing the code I may have reacted a bit harshly. The latest commits were mainly refactorings which allowed the webcomponentsjs `@apply`-shim to work. It's unfortunate that these changes increased the bundle size.

If these refactors added significant architectural advantages other than allowing for `@apply` to work, I understand the changes and I will close this issue.

I do want to reiterate that I think my concern is still relevant. One of the major appeals of lit-html was that it is a super lightweight rendering engine. Judging by it's rapid adoption I expect this code to be used by and served to clients many many times in the future, which means there is real value in keeping it small. Changes that increase the size should require a very compelling reason to do so.

**Comment by Lodin** [comment]
@ruphin I would like to note that since `lit-html` has modular structure you can omit importing `shady-renderer` if you don't need it and use only the core part of lit-html which provides all the benefits you mention but without any polyfill support

**Comment by AndreasGalster** [comment]
They can always remove shady-css as opt-out later, once it's less risky to not support it, e. g. in 1-2 years when part/theme are implemented and majority of sites won't be relying on @apply anymore (old sites still using it from Polymer times)

**Comment by ruphin** [comment]
But I do want to use the `shady-renderer` to support ShadyDOM. Supporting a ShadowDOM polyfill makes sense, because ShadowDOM is a standard, and at some point we can stop using the polyfill and go our happy ways. It's different from supporting the `@apply` shim, because that is not a standard, and it will never be supported.

Even the ShadowDOM polyfill doesn't really 'support' the `@apply` shim anymore. They are only leaving it in because removing it will cause considerable pain for legacy users when they upgrade. This is exactly why I think it's a bad idea to support `@apply` in lit-html now. Once you add support for something it's REALLY hard to remove it, because you'll hinder all the users that have come to rely on it.

Nobody in the lit-html space is relying on `@apply` now (that would be impossible since it literally wasn't supported), and it's a bad idea to start doing so now, because it will only cause pain down the road. You're deliberately building on a deprecated standard if you do.

**Comment by AndreasGalster** [comment]
I am relying on @apply... My old views converted to lit have polymer elements using @apply styles.

Don't fall to conclusions :)

**Comment by ruphin** [comment]
You are relying on `@apply` but not on lit-html. You're trying to move to lit-html and bring `@apply` with you. Right before this change `@apply` literally did not work in lit-html, so it is technically impossible for anyone to have a project currently deployed with lit-html that relies on `@apply` to work.

Your code is still going to rely on `@apply` 1-2 years from now if you're just going to port it to lit-html. Legacy code is not going to magically go away. If moving to another rendering engine isn't enough reason to refactor your code off of deprecated APIs, then what will? This is exactly my point. If we are going to support `@apply` now, people are going to move to lit-html with their `@apply` code. Then when we want to remove it later, we can't because people rely on it.

If you're moving to a different platform, you should expect to have to refactor code that is compatible with that platform. There's no reason why `@apply` should be compatible with this platform.

**Comment by ruphin** [comment]
I'm not asking you to do anything. Who is forcing you to migrate to lit-html? Polymer 3 works just fine doesn't it?

I don't understand how you expect there to be a comfortable migration path for you to move from one platform to another. It's like moving from React to lit-html and expecting there to be JSX support.

**Comment by AndreasGalster** [comment]
That's contradicting what the polymer team has sold polymer since 0.5x though. The point of Polymer (not webcomponents polyfill) was to eliminate the layers that are added sugar over time. Now they are finally removing the sugar and we should completely stay on old Polymer. If that's what it is then this is how the team should have advertised it from the start, because frankly then one could have just stuck with React if the plan was always to let people just abandon future work on long-term projects because the modern approach is not compatible.

This is just the exact opposite of what the team has presented as their vision. If this would be the position of the team now, I'd be incredibly disappointed for being dishonest for years with the people who adopted Polymer despite obvious drawbacks compared to other libraries.

**Comment by abdonrd** [comment]
I mention the same here: https://github.com/Polymer/lit-html/pull/337#issuecomment-391248855

<img width="790" alt="screen shot 2018-06-17 at 10 44 34" src="https://user-images.githubusercontent.com/1007051/41506300-7a6a0e7c-721b-11e8-8315-711415c80970.png">

Without having seen how the implementation is, I prefer drop `@apply` support.
I think new elements with LitElement should avoid deprecated things like `@apply`.

---

The Vaadin team have a nice approach with their [vaadin-themable-mixin](https://github.com/vaadin/vaadin-themable-mixin) to start using parts.

* [::part and ::theme, an ::explainer](https://meowni.ca/posts/part-theme-explainer/)
* [CSS Shadow Parts](https://tabatkins.github.io/specs/css-shadow-parts/)

**Comment by kenchris** [comment]
It's really sad to see that nothing happens spec wise! 

> Unofficial Proposal Draft, 20 January 2017

**Comment by abdonrd** [comment]
@kenchris totally agree!

**Comment by kenchris** [comment]
@justinfagnani do you know about the status of that spec? no works seems to have been done since January last year, and I know this is worrying a lot of people, incl. people from Ionic. What is needed to get this rolling? New editor?

**Comment by ruphin** [comment]
@AndreasGalster I don't understand how not supporting `@apply` is contradictory to the original vision of Polymer.

First of all, Polymer 3 is not the "old Polymer" and lit-html is not the "new Polymer". Polymer 3 is the new Polymer. Lit-html is an independent new project. I invite you to look at the [lit-html project README](https://github.com/Polymer/lit-html#motivation). Small size and standards-compliance are right there. Nowhere is mentioned that one of the goals of lit-html is to allow an easy migration path from Polymer (or from any other framework for that matter).

Second, the promise of removing additional layers over the platform itself is exactly what this project is doing. `@apply` is not the 'platform way' of doing style sharing, it is just a sugar over an original idea that didn't end up making it as a platform specification. In your words, removing additional layers of sugar is exactly the point of Polymer. Removing `@apply` seems perfectly consistent with that.

Look, I understand your point. You want to migrate to the latest thing, but rewriting your stack is a massive pain. I get it. I had a massive Polymer 0.5 application running in production until about a year ago, because it was too much effort to refactor. I too was really disappointed when HTML imports were deprecated, because I my entire codebase depended on it, and I really liked the spec.

But this is partly what we signed up for when we chose Polymer. Targeting future standards is tricky because they can be changed from underneath you at any point. When the Web Component specs changed from v0 to v1, we had to refactor to accommodate those changes. Now that HTML imports have been deprecated, we have to refactor for that. When you wrote code that relied on `@apply` you could have known that there was a chance the specification wouldn't make it, and that Polymer would move towards whatever new standard would appear. To me, this is completely consistent with the original vision. 

The important thing is that you have choice. If you want to write lean standards-based code, you can use lit-html. If you want to retain legacy APIs so you don't have to rewrite your stack, you can use Polymer (3). Both are great choices. Polymer didn't suddenly become 'bad' or 'old' now that another option appeared.

**Comment by web-padawan** [comment]
The only thing `@apply` shim does: splits "mixins" into custom CSS properties

AFAIK, existing draft of the `::part` polyfill does essentially the same. So, as for now, both shims are syntactic sugar on top of the CSS custom properties.

**Comment by notwaldorf** [comment]
@kenchris You're looking at the wrong page. It was already moved to the [csswg](https://drafts.csswg.org/css-shadow-parts-1/), where it's already gone through several rounds of review

**Comment by kenchris** [comment]
@notwaldorf having the wrong link/spec add a note saying that would make a lot of sense. We have done that in other similar cases before

**Comment by jsilvermist** [comment]
If they both use custom properties in the polyfills, is there really that big of a performance difference to polyfills part vs apply?

Is it just waking through nodes for part that makes the polyfill more costly?

**Comment by notwaldorf** [comment]
@kenchris Tab is going to add a redirect to the correct spec.

**Comment by justinfagnani** [comment]
Hi all, Sorry I haven't weighed in here yet.

As @Lodin pointed out, lit-html doesn't support `@apply` directly - it supports various hooks to customize template preparation and instance stamping that let the shady-render module support ShadyCSS. The initial version of those hooks and module didn't fully support ShadyCSS because the complete set of styles stamped into a shadow root were not known at template prep time, but only after a render. This most obviously meant that `@apply` didn't work, but it also meant that the most likely way that `::part` and `::theme` would be polyfilled wouldn't work as well.

But let's break the issues down into a few main points:

1. shady-render got bigger. Yes, this is unfortunate, but it does a lot more in terms of letting templates be rewritten, and in addition to `@apply`, should support one-time bindings in `<style>` and future CSS Shadow Part work.

    Also, we have to dig and and do some performance and size optimization work here. We may be able to bring the size down.

2. `@apply` _is_ important to us. Yes, it's a dead-end, but it's used and critically it's not just an implementation detail of a component, it's basically part of it's public API. So if we want to enable migration from Polymer to LitElement we have to allow producing and consuming `@apply` mixin values. There is a lot of code both internally and externally that we want to help migrate as soon and as painlessly as possible, so `@apply` support really isn't optional here.

3. `::part` and `::theme` have similar polyfill complexities as `@apply`. Just because they both involve coordination across levels of the tree, they end up requiring knowledge of the tree relationship between setting and declaring elements. We have tried to brainstorm some ideas around a subset of Shadow Parts, or something that essentially compiles to Shadow Parts, to try to avoid that complexity, but honestly we don't have anything amazing yet, and just might not. So we'll probably need the same facilities that the latest updates added.

4. It's possible to write an alternate `render()` function that doesn't do all the work to support instance-time template modification for `@apply` or `::part` and is smaller. It'll mostly be the old shady-render. If the extra size is that bad for projects that also have to have ShadyCSS support, I can help someone write that. I'm not sure we want to have to support

5. Alternatively, maybe we can figure out a way to only load shady-render when the polyfill needs to kick in, so you only pay the size if you 1) need the polyfill or 2) actually use `@apply`. In this case Chrome, Safari and soon Firefox wouldn't pay any penalty at all. I'm not sure what this might look like yet.

@ruphin I understand your sentiment for wanting the smallest possible library that fits your requirements, but there are a lot of permutation on requirements out there, and we won't be able to make the smallest thing for each permutation exactly. Given our other priorities and commitments (and that we're going to almost certainly need this approach for Shadow Parts), we really can't remove support now. Thanks for sparking this discussion, but I'm going to close the issue as it's not feasible.

**Comment by ruphin** [comment]
Thanks for the detailed reply. As I mentioned before I was probably a bit harsh on the topic, and I mainly wanted to express concern about performance/size tradeoffs. I understand that you are working within certain constraints and choices have to be made. Given the likelihood of requiring these changes for ::part and ::theme I think this choice is very valid.

I'm heavily invested in lit-html at this point, with nearly all my projects depending on it. I spend significant time professionally advocating the use of lit-html during consulting gigs and speaking engagements. If I'm being overly critical, it's only because I care deeply about it's success. :)

I'm overall super happy about how the project is run, and just wanted you to know I'm extremely grateful for your continued efforts.
